### ISSUE TITLE
Addon Docs: decorated components break control reactivity in addon-docs props tables

### ISSUE THREAD

**Issue by Tri101** [storybookjs/storybook#11244]
**Description**
I'm in the process of migrating knobs to controls, and it appears that props defined as controls are not being bound properly in the addon-docs tab.  The props are not being bound in the primary example.

I'm using the latest beta (6.0.0-beta.37).  The component and controls function as expected in the canvas tab with no console warning. 

When I pass in my story args as data props, it fixes the data binding/reactivity in the docs tab, but breaks reactivity of the controls on the canvas tab.

I am also using a global decorator to do some theme switching, and a simple template decorator defined at the story level.  It appears that passing the decorated component to the DocsPage is what could be causing my issue.  If I remove the decorators

**To Reproduce**
Given a Vue component story with props defined as controls,
And a simple template decorator
1. Go to the addon-docs tab 
2. Interact with a control in the props table
3. Controls should update the props in the component (expected), controls do not update props in the component (actual).

**Expected behavior**
Given some props defined as controls, when I navigate to the addon-docs tab, then I expect the the controls in the props table to update the props in the docs page primary example. 

**Screenshots**
![test-canvas](https://user-images.githubusercontent.com/7559656/85139000-c820c380-b211-11ea-833d-3896e5d72f38.png)
![test-docs](https://user-images.githubusercontent.com/7559656/85139002-c8b95a00-b211-11ea-97c8-9f3a9d77c696.png)

**Code snippets**
```
// component.vue

<template>
	<div>{{ text }}</div>
</template>
<script>
export default {
	props: {
		text: {
			default: 'test',
			type: String
		}
	}
};
</script>



// component.story.js

import MyComponent from './component.vue';

const templateDecorator = () => ({
	template: `
<div style="background-color: gray;"><story/></div>
	`
});

export const Test = (args) => ({
	components: { MyComponent },
	props: Object.keys(args),
	template: `
<my-component :text="text"></my-component>
	`
});

Test.args = {
	text: 'this is only a test'
};

Test.decorators = [
        templateDecorator
]

export default {
	component: MyComponent,
	title: 'Components/MyComponent'
};

```

**System:**

  System:
    OS: macOS Mojave 10.14.6
    CPU: (8) x64 Intel(R) Core(TM) i7-4960HQ CPU @ 2.60GHz
  Binaries:
    Node: 10.16.0 - ~/.nvm/versions/node/v10.16.0/bin/node
    npm: 6.9.0 - ~/.nvm/versions/node/v10.16.0/bin/npm
  Browsers:
    Safari: 12.1.2
  npmPackages:
    @storybook/addon-actions: 6.0.0-beta.37 => 6.0.0-beta.37 
    @storybook/addon-backgrounds: 6.0.0-beta.37 => 6.0.0-beta.37 
    @storybook/addon-controls: 6.0.0-beta.37 => 6.0.0-beta.37 
    @storybook/addon-docs: 6.0.0-beta.37 => 6.0.0-beta.37 
    @storybook/addon-storysource: 6.0.0-beta.37 => 6.0.0-beta.37 
    @storybook/vue: 6.0.0-beta.37 => 6.0.0-beta.37

**Comment by shilman** [comment]
Do you have a repro repo? This all looks ok to me, and is working properly in the storybook monorepo: https://github.com/storybookjs/storybook/blob/next/examples/vue-kitchen-sink/src/stories/addon-controls.stories.js

**Comment by saschafuchs** [comment]
I have a similar problem with Nuxt in combination with vue-i18n. 

i18n I have included as a decorator in the `preview.js`

```js 
addDecorator(() => ({
  template: '<story/>',
  i18n
}))
```

When the translation is included, I can enter values in the props, but they are no longer transferred to the component. If I deactivate the decorator, it works as desired. But this is not the desired way, because I need the translations in the components.

**Comment by Tri101** [comment]
Same. While putting together me repro I discovered that my issue was related to my use of decorators as well.

Seems that the props are being ignored by the decorators.  I'm using both a global decorator to apply theming to all my stories, as well as templateDecorators defined at the story level (for certain stories).  I'm looking into a way to have addon-docs consume a decorated component, and still have controls in the props table be reactive, but I'm not sure if it's possible.

Still digging...

**Comment by Tri101** [comment]
Link to the repro: https://github.com/Tri101/11244-addon-docs-decorators-issue

**Comment by guilhermewaess** [comment]
Same problem here, when using decorators, the controls in docs page stopped to work. Without decorators works fine.
I'm using .RC-3

**Comment by shilman** [comment]
@guilhermewaess are you also using vue?

**Comment by shilman** [comment]
I did a little digging and found that there are two potential culprits:

### The Docs-specific hooks that listens for changes to Args

https://github.com/storybookjs/storybook/blob/next/addons/docs/src/blocks/Props.tsx#L46-L67

### The inline rendering of vue elements 

https://github.com/storybookjs/storybook/blob/next/addons/docs/src/frameworks/vue/config.tsx#L10-L13

I put a `console.log` in both places in `node_modules` and it appears that it's a problem with the inline rendering.

So I tried adding the following to `Button.stories.js`

```js
  parameters: {
    docs: { inlineStories: false },
  },
```

At that point the button renders properly, but it's still not reactive in Docs.

I also made a change to the story that didn't help:

```js
const ButtonStory = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  ...
```

Posting this all for @yannbf who wants to dig into this more. Cc @Aaron-Pool @backbone87 who might have some ideas about what could be going on here.

**Comment by Aaron-Pool** [comment]
@shilman are decorators framework specific, or framework agnostic? The template decorator in the original comment looks kinda like a Vue component, so I was just curious if there's any per framework implementation going on with the decorator feature.

**Comment by shilman** [comment]
@Aaron-Pool good question. it looks like there's some vue-specific decorator logic here:

https://github.com/storybookjs/storybook/blob/next/app/vue/src/client/preview/index.ts#L76-L104

**Comment by guilhermewaess** [comment]
> @guilhermewaess are you also using vue?

Nope, react.

**Comment by Aaron-Pool** [comment]
@shilman To be totally honest, based on that code I'm not sure how the target story is getting even the initial props value, much less reacting to them. I'll try to look at this later, just about to start my work day.

**Comment by shilman** [comment]
@Aaron-Pool it's a little magic. @backbone87 and I modified the vue code to store the args in the story's reactive store here:

https://github.com/storybookjs/storybook/blob/next/app/vue/src/client/preview/render.ts#L57

and then it gets passed to the story as props here

https://github.com/storybookjs/storybook/blob/next/app/vue/src/client/preview/index.ts#L58

**Comment by Aaron-Pool** [comment]
@shilman yeah, but the way I'm reading it, that's binding them as props to the root node of the story decorator, but in the original example the root node of the decorator is not the actually the story component, but a wrapper div applying styles. So I don't see how the props are getting to the final story component, am I being an idiot? 🤷‍♂️

**Comment by shilman** [comment]
@Aaron-Pool that vue knowledge is above my pay grade, but that could be the problem? @backbone87 ?

**Comment by backbone87** [comment]
ok short explanation how it worked with sb4/5 knobs:
- each story & decorator component gets normalized with Vue.extend
- then each one gets wrapped with a functional component (lets call this the wrapper)
- the wrapper has 2 responibilities:
  - collect prop default values from the inner wrapper (if any) and from the wrapped components options https://github.com/storybookjs/storybook/blob/next/app/vue/src/client/preview/index.ts#L50
  - inject knob values stored on the $root (this also includes control values in sb6) into the props of the wrapped component https://github.com/storybookjs/storybook/blob/next/app/vue/src/client/preview/index.ts#L58

so the result of this whole decorate/prepare is:
https://github.com/storybookjs/storybook/blob/next/app/vue/src/client/preview/render.ts#L36
this is a component (**not** component instance), which contains the collected default prop values (and therefore current knob values) in the property `STORYBOOK_VALUES`

if $root does not have a component set yet (or we force rerender) then we set this result as the component https://github.com/storybookjs/storybook/blob/next/app/vue/src/client/preview/render.ts#L53

afterwards we update the $root (component instance) property `STORYBOOK_VALUES` with the collected values.

basically on each knob change we do the whole wrapping/collecting (to get the current knob values stored in the prop default values) and then try to reuse the existing component instance (if the story didnt change) to avoid replacing the whole component tree and let vue reactivity do the update for us.

**Comment by shilman** [comment]
Jeepers creepers!! I just released https://github.com/storybookjs/storybook/releases/tag/v6.0.0-rc.9 containing PR #11594 that references this issue. Upgrade today to try it out!

You can find this prerelease on the `@next` NPM tag.

Closing this issue. Please re-open if you think there's still more to do.
